{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All tests passed successfully for examples in set1\n",
      "All tests passed successfully for examples in set2\n",
      "Question1:  85000\n",
      "Question2:  104266\n",
      "Question3:  15291\n",
      "Question4: How many times does the string GGCGCGGTGGCTCACGCCTGTAAT, which is derived from a human Alu sequence, occur with up to 2 substitutions in the excerpt of human chromosome 1?\n",
      "2\n",
      "check with naive_2mm validated\n",
      "Question5: how many total index hits are there when searching for occurrences of GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the excerpt of human chromosome 1?\n",
      "3\n",
      "occurrences, num_index_hits: [0, 14] 3\n",
      "All tests passed successfully for examples in set3\n",
      "Question 6: how many total index hits are there when searching for GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the excerpt of human chromosome 1?\n",
      "[36918, 116447] 3\n"
     ]
    }
   ],
   "source": [
    "def naive_with_counts(p, t):\n",
    "    occurrences = []\n",
    "    num_char_comp = 0\n",
    "    num_aligments_tried = 0\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        num_aligments_tried += 1\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            num_char_comp += 1\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences, num_char_comp, num_aligments_tried\n",
    "\n",
    "def example_1_1():\n",
    "    p = 'word'\n",
    "    t = 'there would have been a time for such a word'\n",
    "    #print(naive_with_counts(p, t))\n",
    "    assert naive_with_counts(p, t) == ([40], 46, 41)\n",
    "\n",
    "def example_1_2():\n",
    "    p = 'needle'\n",
    "    t = 'needle need noodle needle'\n",
    "    #print(naive_with_counts(p, t))\n",
    "    assert naive_with_counts(p, t) == ([0, 19], 35, 20)\n",
    "\n",
    "def boyer_moore_with_counts(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching. p=pattern, t=text, p_bm=BoyerMoore object for p \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    num_char_comp = 0\n",
    "    num_aligments_tried = 0\n",
    "    \n",
    "    while i < len(t) - len(p) + 1:\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        num_aligments_tried += 1\n",
    "\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            num_char_comp += 1\n",
    "            if p[j] != t[i+j]:    \n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "        \n",
    "    return occurrences, num_char_comp, num_aligments_tried\n",
    "\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "def example_2_1():\n",
    "    from bm_preproc import BoyerMoore\n",
    "    p = 'word'\n",
    "    t = 'there would have been a time for such a word'\n",
    "    lowercase_alphabet = 'abcdefghijklmnopqrstuvwxyz '\n",
    "    p_bm = BoyerMoore(p, lowercase_alphabet)\n",
    "    #print(boyer_moore_with_counts(p, p_bm, t))\n",
    "    assert boyer_moore_with_counts(p, p_bm, t) == ([40], 15, 12)\n",
    "\n",
    "def example_2_2():\n",
    "    from bm_preproc import BoyerMoore\n",
    "    p = 'needle'\n",
    "    t = 'needle need noodle needle'\n",
    "    lowercase_alphabet = 'abcdefghijklmnopqrstuvwxyz '\n",
    "    p_bm = BoyerMoore(p, lowercase_alphabet)\n",
    "    #print(boyer_moore_with_counts(p, p_bm, t))\n",
    "    assert boyer_moore_with_counts(p, p_bm, t) == ([0, 19], 18, 5)\n",
    "\n",
    "def question1_and_2():\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'\n",
    "    reads, qualities = readFastq('chr1.GRCh38.excerpt.fasta')\n",
    "    assert len(reads) == len(qualities)\n",
    "\n",
    "    total_char_comp = 0\n",
    "    total_align_comp = 0\n",
    "\n",
    "    for t in reads:\n",
    "        occurrences, num_char_comp, num_aligments_tried = naive_with_counts(p, t)\n",
    "        total_char_comp += num_char_comp\n",
    "        total_align_comp += num_aligments_tried\n",
    "\n",
    "    \"\"\"How many alignments does the naive exact matching algorithm try when matching the string \n",
    "    GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG (derived from human Alu sequences) to the \n",
    "    excerpt of human chromosome 1? (Don't consider reverse complements.)\"\"\"\n",
    "    print 'Question1: ', total_align_comp\n",
    "\n",
    "    \"\"\"How many character comparisons does the naive exact matching algorithm try when matching \n",
    "    the string GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG (derived from human Alu sequences) \n",
    "    to the excerpt of human chromosome 1? (Don't consider reverse complements.)\"\"\"\n",
    "    print 'Question2: ', total_char_comp\n",
    "    \n",
    "def question3():\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'\n",
    "    reads, qualities = readFastq('chr1.GRCh38.excerpt.fasta')\n",
    "    assert len(reads) == len(qualities)\n",
    "\n",
    "    total_char_comp = 0\n",
    "    total_align_comp = 0\n",
    "\n",
    "    from bm_preproc import BoyerMoore\n",
    "    p_bm = BoyerMoore(p)\n",
    "    for t in reads:\n",
    "        occurrences, num_char_comp, num_aligments_tried = boyer_moore_with_counts(p, p_bm, t)\n",
    "        total_char_comp += num_char_comp\n",
    "        total_align_comp += num_aligments_tried\n",
    "\n",
    "    \"\"\"How many alignments does Boyer-Moore try when matching the string \n",
    "    GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG (derived from human Alu sequences) \n",
    "    to the excerpt of human chromosome 1? (Don't consider reverse complements.)\"\"\"\n",
    "    print 'Question3: ', total_align_comp\n",
    "    #print 'Question3--: ', total_char_comp\n",
    "\n",
    "class Index(object):\n",
    "    def __init__(self, t, k):\n",
    "        ''' Create index from all substrings of size 'length' '''\n",
    "        self.k = k  # k-mer length (k)\n",
    "        self.index = []\n",
    "        for i in range(len(t) - k + 1):  # for each k-mer\n",
    "            self.index.append((t[i:i+k], i))  # add (k-mer, offset) pair\n",
    "        self.index.sort()  # alphabetize by k-mer\n",
    "    \n",
    "    def query(self, p):\n",
    "        import bisect\n",
    "        ''' Return index hits for first k-mer of P '''\n",
    "        kmer = p[:self.k]  # query with first k-mer\n",
    "        i = bisect.bisect_left(self.index, (kmer, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != kmer:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits\n",
    "\n",
    "def question4():\n",
    "    \"\"\"Write a function that, given a length-24 pattern P and given an Index object built on 8-mers, finds all approximate occurrences of P within T with up to 2 mismatches. \n",
    "    Insertions and deletions are not allowed. Don't consider any reverse complements.\n",
    "    How many times does the string GGCGCGGTGGCTCACGCCTGTAAT, which is derived from a human Alu sequence, occur with up to 2 substitutions in the excerpt of human \n",
    "    chromosome 1? (Don't consider reverse complements here.)\n",
    "    - Hint 1: Multiple index hits might direct you to the same match multiple times, but be careful not to count a match more than once.\n",
    "    - Hint 2: You can check your work by comparing the output of your new function to that of the naive_2mm function implemented in the previous module.\"\"\"\n",
    "\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n",
    "\n",
    "    #number of mistmatches = 2, so we need to split into 3 (2+1)\n",
    "    mistmatches_allowed = 2\n",
    "    num_segments_required = mistmatches_allowed + 1\n",
    "    k_mer_size = 8\n",
    "    pattern_size = 24\n",
    "\n",
    "    reads, qualities = readFastq('chr1.GRCh38.excerpt.fasta')\n",
    "    consolidated_read = ''.join([read for read in reads])\n",
    "\n",
    "    index = Index(consolidated_read, k_mer_size)\n",
    "    p_segments = [ p[i*k_mer_size:i*k_mer_size+k_mer_size] for i in range(num_segments_required) ] # = ['GGCGCGGT', 'GGCTCACG', 'CCTGTAAT']\n",
    "    #print p_segments\n",
    "\n",
    "    hits_per_segment = {} # not used\n",
    "    hit_lists = []\n",
    "    i = 0\n",
    "    index_hits = 0\n",
    "    for segment in p_segments:\n",
    "        hits = index.query(segment)\n",
    "        index_hits += 1\n",
    "        if len(hits) > 0:\n",
    "            #print segment, hits\n",
    "            hits_per_segment[segment] = hits\n",
    "            hit_lists.append(set([hit - i*k_mer_size for hit in hits]))  # we keep the starting point of the pattern in our results set for easy comparison (verification means equality)\n",
    "                                                                         # and we store them in set for easy equality check by means of intersecting the sets.\n",
    "        i += 1\n",
    "    #print hits_per_segment\n",
    "    #print hit_lists\n",
    "\n",
    "        \n",
    "    results = []\n",
    "    for i in range(len(hit_lists)):\n",
    "        # WARN: this bit here only works for numbr of mismatches = 2 as we intersect two sets for any values of mistmatches_allowed, it would lead to invalid results otherwise\n",
    "        intersect = hit_lists[i].intersection(hit_lists[(i+1)%len(hit_lists)])\n",
    "        if len(intersect) > 0 : \n",
    "            for item in intersect:\n",
    "                results.append(item)\n",
    "    print len(results)\n",
    "    return sorted(results), index_hits\n",
    "\n",
    "# Copy paste from homework1.py\n",
    "def naive_2mm(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        mismatches = 0\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                mismatches += 1\n",
    "                if mismatches > 2:\n",
    "                    match = False\n",
    "                    break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences\n",
    "\n",
    "def question4_check():\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n",
    "    reads, qualities = readFastq('chr1.GRCh38.excerpt.fasta')\n",
    "    t = ''.join([read for read in reads])\n",
    "    return naive_2mm(p, t)\n",
    "\n",
    "class SubseqIndex(object):\n",
    "    \"\"\" Holds a subsequence index for a text T \"\"\"\n",
    "    \n",
    "    def __init__(self, t, k, ival):\n",
    "        \"\"\" Create index from all subsequences consisting of k characters\n",
    "            spaced ival positions apart.  E.g., SubseqIndex(\"ATAT\", 2, 2)\n",
    "            extracts (\"AA\", 0) and (\"TT\", 1). \"\"\"\n",
    "        self.k = k  # num characters per subsequence extracted\n",
    "        self.ival = ival  # space between them; 1=adjacent, 2=every other, etc\n",
    "        self.index = []\n",
    "        self.dictIndex = {}\n",
    "        self.span = 1 + ival * (k - 1)\n",
    "        for i in range(len(t) - self.span + 1):  # for each subseq\n",
    "            self.index.append((t[i:i+self.span:self.ival], i))  # add (subseq, offset)\n",
    "            if self.dictIndex.has_key(t[i:i+self.span:self.ival]):\n",
    "                self.dictIndex[t[i:i+self.span:self.ival]].append(i)\n",
    "            else:\n",
    "                self.dictIndex[t[i:i+self.span:self.ival]] = [i]\n",
    "        self.index.sort()  # alphabetize by subseq\n",
    "        #print \"built index\", self.index\n",
    "        #print \"dict index\", self.dictIndex\n",
    "    \n",
    "    def get_subseq(self, p):\n",
    "        return [p[i:i+self.span:self.ival] for i in range(len(p) - self.span + 1)]  # add (subseq, offset)\n",
    "\n",
    "    def queryDictIndex(self, p):\n",
    "        subseq = p[:self.span:self.ival]  # query with first subseq\n",
    "        if len(subseq) != self.k:\n",
    "            return []\n",
    "\n",
    "        #print \"input is: '{0}' (length={1}). Querying for: '{2}' (length= {3})\".format(p, len(p), subseq, len(subseq))\n",
    "        \n",
    "        #print \"dict index\", self.dictIndex\n",
    "        if subseq in self.dictIndex.keys():\n",
    "            return self.dictIndex[subseq]\n",
    "        else:\n",
    "           return []\n",
    "\n",
    "    def query(self, p):\n",
    "        \"\"\" Return index hits for first subseq of p \"\"\"\n",
    "        import bisect\n",
    "        subseq = p[:self.span:self.ival]  # query with first subseq\n",
    "        #print \"input is: '{0}' (length={1}). Querying for: '{2}' (length= {3})\".format(p, len(p), subseq, len(subseq))\n",
    "        i = bisect.bisect_left(self.index, (subseq, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != subseq:\n",
    "                print \"'\", self.index[i][0], \"' != '\", subseq, \"'\"\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits\n",
    "\n",
    "def query_subseq(p, t, subseq_ind):\n",
    "    \"\"\"Write a function that, given a length-24 pattern P and given a SubseqIndex object built with k = 8 and ival = 3, \n",
    "    finds all approximate occurrences of P within T with up to 2 mismatches.\"\"\"\n",
    "    #number of mistmatches = 2, so we need to split into 3 (2+1)\n",
    "    mistmatches_allowed = 2\n",
    "    num_segments_required = mistmatches_allowed + 1\n",
    "    \n",
    "    pattern_size = 24\n",
    "\n",
    "    reads, qualities = readFastq('chr1.GRCh38.excerpt.fasta')\n",
    "    consolidated_read = ''.join([read for read in reads])\n",
    "\n",
    "    #p_segments = [ p[i:] for i in range(0,  pattern_size / subseq_ind.ival) ]\n",
    "    p_segments = []\n",
    "    for i in range(0,  pattern_size):\n",
    "        if len(p[i::subseq_ind.ival]) == subseq_ind.k:\n",
    "            p_segments.append(p[i:])\n",
    "        else:\n",
    "            break\n",
    "    #print p_segments\n",
    "\n",
    "    #print [subseq_ind.get_subseq(segment) for segment in p_segments]\n",
    "    checksize = reduce(lambda x,y : x and y, [map(lambda x: len(x)==8, subseq_ind.get_subseq(segment)) for segment in p_segments])[0]\n",
    "    assert checksize\n",
    "    \n",
    "    hits_lists = []\n",
    "    hits_per_segment = {}\n",
    "    i = 0\n",
    "    index_hits = 0\n",
    "    segment_number = 0\n",
    "    for segment in p_segments:\n",
    "        \n",
    "        #hits = subseq_ind.query(segment)\n",
    "        hits = subseq_ind.queryDictIndex(segment)\n",
    "        i += 1\n",
    "        #print \"hits: \", hits\n",
    "        index_hits += 1\n",
    "        if len(hits) > 0:\n",
    "            #print segment, hits\n",
    "            hits_lists.append(hits)\n",
    "            hits_per_segment[segment_number] = set([hit-segment_number for hit in hits])\n",
    "        segment_number += 1\n",
    "    #print hits_per_segment\n",
    "\n",
    "    #reduce:\n",
    "    reduced_hits = set()\n",
    "    for i in range(len(hits_per_segment)):\n",
    "        intersect = hits_per_segment[i].intersection(hits_per_segment[(i+1)%len(hits_per_segment)])\n",
    "        if len(intersect) > 0 : \n",
    "            for item in intersect:\n",
    "                reduced_hits.add(item)\n",
    "    \n",
    "    return sorted(reduced_hits), index_hits\n",
    "\n",
    "\n",
    "def example_3_1():\n",
    "    t = 'to-morrow and to-morrow and to-morrow creeps in this petty pace'\n",
    "    p = 'to-morrow and to-morrow '\n",
    "    subseq_ind = SubseqIndex(t, 8, 3)\n",
    "    occurrences, num_index_hits = query_subseq(p, t, subseq_ind)\n",
    "    print \"occurrences, num_index_hits:\", occurrences, num_index_hits\n",
    "    #assert occurrences == [0, 14]\n",
    "    #this seems invalid: \n",
    "    #assert num_index_hits == 6\n",
    "\n",
    "\n",
    "def question6():\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n",
    "\n",
    "    #number of mistmatches = 2, so we need to split into 3 (2+1)\n",
    "    mistmatches_allowed = 2\n",
    "    num_segments_required = mistmatches_allowed + 1\n",
    "    k_mer_size = 8\n",
    "    pattern_size = 24\n",
    "    ival = 3\n",
    "\n",
    "    reads, qualities = readFastq('chr1.GRCh38.excerpt.fasta')\n",
    "    t = ''.join([read for read in reads])\n",
    "\n",
    "    subseq_ind = SubseqIndex(t, k_mer_size, ival)\n",
    "    return query_subseq(p, t, subseq_ind)\n",
    "\n",
    "def main():\n",
    "    example_1_1()\n",
    "    example_1_2()\n",
    "    print \"All tests passed successfully for examples in set1\"\n",
    "    example_2_1()\n",
    "    example_2_2()\n",
    "    print \"All tests passed successfully for examples in set2\"\n",
    "    question1_and_2()\n",
    "    question3()\n",
    "    print \"Question4: How many times does the string GGCGCGGTGGCTCACGCCTGTAAT, which is derived from a human Alu sequence, occur with up to 2 substitutions in the excerpt of human chromosome 1?\"\n",
    "    res, num_index_hits = question4()\n",
    "    check_res = question4_check()\n",
    "    assert res == check_res\n",
    "    print \"check with naive_2mm validated\"\n",
    "    print \"Question5: how many total index hits are there when searching for occurrences of GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the excerpt of human chromosome 1?\"\n",
    "    print num_index_hits\n",
    "    example_3_1()\n",
    "    print \"All tests passed successfully for examples in set3\"\n",
    "    print \"Question 6: how many total index hits are there when searching for GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the excerpt of human chromosome 1?\"\n",
    "    occurrences, num_index_hits = question6()\n",
    "    print occurrences, num_index_hits\n",
    "   \n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
